src/mscorlib/src/System/String.cs

```diff
@@ -3338,34 +3338,66 @@ public static String Concat(Object arg0, Object arg1, Object arg2, Object arg3,
         }
 
         [System.Security.SecuritySafeCritical]
-        public static String Concat(params Object[] args) {
-            if (args==null) {
+        public static string Concat(params object[] args)
+        {
+            if (args == null)
+            {
                 throw new ArgumentNullException("args");
             }
             Contract.Ensures(Contract.Result<String>() != null);
             Contract.EndContractBlock();
-    
-            String[] sArgs = new String[args.Length];
-            int totalLength=0;
-            
-            for (int i=0; i<args.Length; i++) {
+
+            // We need to get an intermediary string array
+            // to fill with each of the args' ToString(),
+            // and then just concat that in one operation.
+
+            // This way we avoid any intermediary string representations,
+            // or buffer resizing if we use StringBuilder (although the
+            // latter case is partially alleviated due to StringBuilder's
+            // linked-list style implementation)
+
+            var strings = new string[args.Length];
+
+            long totalLengthLong = 0L;
+
+            for (int i = 0; i < args.Length; i++)
+            {
                 object value = args[i];
-                sArgs[i] = ((value==null)?(String.Empty):(value.ToString()));
-                if (sArgs[i] == null) sArgs[i] = String.Empty; // value.ToString() above could have returned null
-                totalLength += sArgs[i].Length;
-                // check for overflow
-                if (totalLength < 0) {
-                    throw new OutOfMemoryException();
-                }
+
+                string toString = value?.ToString() ?? string.Empty; // We need to handle both the cases when value or value.ToString() is null
+                strings[i] = toString;
+
+                totalLengthLong += toString.Length;
+            }
+
+            if (totalLengthLong > int.MaxValue)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            int totalLength = (int)totalLengthLong;
+
+            // If all of the ToStrings are null/empty, just return string.Empty
+            if (totalLength == 0)
+            {
+                return string.Empty;
             }
 
             string result = FastAllocateString(totalLength);
-            int currPos = 0;
-            for (int i = 0; i < sArgs.Length; i++)
+            int position = 0; // How many characters we've copied so far
+
+            // It's important that this is this is strings.Length
+            // and not args.Length, since RyuJIT is currently unable
+            // to eliminate range checks for the latter
```

Comment:
This may not be worth the comment. It is a well-known fact that you get best performance in for-loops over arrays when the limit is length of the array that is being iterated on, and not some other derived value.